### PR TITLE
Tests and improved docs for unordered groups

### PR DESCRIPTION
**Pull request by igordejanovic**
refs #358

## Code review checklist

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] Changelog(s) has/have been updated, as needed (see `CHANGELOG.md`, no need
      to update for typo fixes and such).

[commit messages]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: docs/grammar.md [hunk 1]
```diff
@@ -262,8 +262,8 @@ Now, an ordered choice in the parentheses is optional.
 
 * **Unordered group** is a special kind of a sequence. Syntactically it is
   similar to a repetition. It is specified by the `#` operator and must be
-  applied to sequences. This operator will match each element of the sequence in
-  an arbitrary order:
+  applied to either sequences or ordered choices. This operator will match each
+  element of the sequence in an arbitrary order:
```
Comment (goto40) [inline docs/grammar.md]: ...will match each element of the sequence +"or the ordered choice"  in an arbitrary order:

**Comment by goto40** [comment]
very good. The addition about ordered choices makes the thing clear.

One more thing: We could add more infos on optional parts of the unordered group:
 * optional parts (e.g. `x ?= 'A'`  work as you intuitively expect: they are accepted in an arbitrary order
 * optional parts are optional - but if all of them are missing and there is no non-optional part, we have a miss (is this correct?)

I am playing with an experimental for-fun-version of parts of textx in c++. There I learn a lot about the internals of textx/arpeggio. I added a unittest on optional parts of unordered groups as I understood it (from the python code): https://github.com/goto40/textx-cpp/blob/master/test/arpeggio.t.cpp#L399

File: examples/pyFlies/pyflies.tx [hunk 2]
```diff
@@ -165,7 +165,7 @@ Block:
 ;
 
 TestInstance:
-  'test' type=[TestType] trials=INT (practice?="practice"|randomize?="randomize")#
+  'test' type=[TestType] trials=INT (practice?="practice" randomize?="randomize")#
```
Comment (goto40) [inline examples/pyFlies/pyflies.tx:168]: "Like"

**Review by goto40 — APPROVED** [review]
LGTM

**Comment by igordejanovic** [comment]
@goto40 
> optional parts are optional - but if all of them are missing and there is no non-optional part, we have a miss (is this correct?)

No, it works as expected. If all elements are optional then even if there is no match of any of the optional parts the whole match still succeeds. At least that's how it should work and a quick test confirms it. The only issue I noticed is that if the whole unordered match is contained in a single common rule than a reference to that rule will return `None` in case all optionals are not matched. Is this "the correct" behavior I don't know. It can be either this or to return an object where all attributes are None. We can investigate this as a separate issue if needed.

> I am playing with an experimental for-fun-version of parts of textx in c++. There I learn a lot about the internals of textx/arpeggio. I added a unittest on optional parts of unordered groups as I understood it (from the python code): https://github.com/goto40/textx-cpp/blob/master/test/arpeggio.t.cpp#L399

Hey, that is excellent news! I look forward for the c++ implementation of textX :)

**Comment by igordejanovic** [comment]
@goto40 Just to clear a bit on my previous statement 

> if there is no match of any of the optional parts the whole match still succeeds

What I mean is that the whole match still succeeds as it is considered optional as a whole.

**Pull request merged**